### PR TITLE
[POAE7-2863] optimize countSetBits and allocBuffer, and close march=native which cause coredump in double*double test

### PR DESCRIPTION
**Pull request by Ma-Jian1**
### What changes were proposed in this pull request?
1. close march=native which cause coredump in double*double test
2. use the attribute to optimize countSetBits, and reduce isBitSet as few as possible
3. **remove `memset` in allocBuffer**, fix offset[0] not init
4. open enable_vectorize for the test
5. fix `reallocate()` for smaller memory

performance detail: [the first version](https://github.com/intel/BDTK/pull/331)

```bash
============================================================================
/workspace/code/BDTK/cpp/src/cider-velox/benchmark/expression/ArithmeticAndComparison.cpprelative  time/iter  iters/s
============================================================================
velox                                                         3.87s  258.34m
nextgen                                           99.97%      3.87s  258.25m
----------------------------------------------------------------------------
nextgen____base                                               3.95s  253.38m
nextgenAVX2                                      412.96%   955.70ms     1.05
nextgenAVX512                                    395.98%   996.68ms     1.00
specifialize1                                   1489.97%   264.88ms     3.78
specifialize2                                   1014.06%   389.19ms     2.57
specifialize3                                    247.55%      1.59s  627.25m
----------------------------------------------------------------------------
mulI8Velox_________Base                                       3.88s  257.47m
mulI8NextGen                                      99.08%      3.92s  255.09m
mulI8NextGen_______Base                                       3.89s  257.23m
mulI8NextgenAVX2                                 400.23%   971.32ms     1.03
mulI8NextgenAVX512                               380.05%      1.02s  977.62m
mulI8NextgenAllOpt                               408.51%   951.63ms     1.05
mulI8Compile                                                44.40ms    22.52
----------------------------------------------------------------------------
mulI8DeepVelox_________Base                                   9.05s  110.44m
mulI8DeepNextGen                                 215.33%      4.21s  237.81m
mulI8DeepNextGen_______Base                                   4.19s  238.66m
mulI8DeepNextgenAVX2                             344.50%      1.22s  822.18m
mulI8DeepNextgenAVX512                           349.47%      1.20s  834.03m
mulI8DeepNextgenAllOpt                           405.73%      1.03s  968.31m
mulI8DeepCompile                                            45.79ms    21.84
----------------------------------------------------------------------------
mulI16Velox_________Base                                      3.29s  303.89m
mulI16NextGen                                     82.43%      3.99s  250.49m
mulI16NextGen_______Base                                      4.15s  240.69m
mulI16NextgenAVX2                                358.26%      1.16s  862.32m
mulI16NextgenAVX512                              355.23%      1.17s  855.01m
mulI16NextgenAllOpt                              345.43%      1.20s  831.43m
mulI16Compile                                               44.79ms    22.33
----------------------------------------------------------------------------
mulI32Velox_________Base                                      3.10s  322.21m
mulI32NextGen                                     74.08%      4.19s  238.69m
mulI32NextGen_______Base                                      4.13s  242.26m
mulI32NextgenAVX2                                327.09%      1.26s  792.40m
mulI32NextgenAVX512                              314.96%      1.31s  763.00m
mulI32NextgenAllOpt                              367.77%      1.12s  890.94m
mulI32Compile                                               45.29ms    22.08
----------------------------------------------------------------------------
mulI64Velox_________Base                                      3.15s  317.61m
mulI64NextGen                                     76.61%      4.11s  243.31m
mulI64NextGen_______Base                                      4.12s  242.50m
mulI64NextgenAVX2                                197.17%      2.09s  478.13m
mulI64NextgenAVX512                              239.18%      1.72s  580.01m
mulI64NextgenAllOpt                              232.73%      1.77s  564.37m
mulI64Compile                                               45.91ms    21.78
----------------------------------------------------------------------------
mulDoubleVelox_________Base                                   3.12s  320.42m
mulDoubleNextGen                                  66.40%      4.70s  212.76m
mulDoubleNextGen_______Base                                   4.68s  213.72m
mulDoubleNextgenAVX2                             260.14%      1.80s  555.96m
mulDoubleNextgenAVX512                           280.01%      1.67s  598.44m
mulDoubleNextgenAllOpt                           279.25%      1.68s  596.81m
mulDoubleCompile                                            45.98ms    21.75
----------------------------------------------------------------------------
mulDoubleSameColumnVelox_________Base                         3.28s  305.10m
mulDoubleSameColumnNextGen                        79.29%      4.13s  241.92m
mulDoubleSameColumnNextGen_______Base                         4.20s  238.15m
mulDoubleSameColumnNextgenAVX2                   319.60%      1.31s  761.14m
mulDoubleSameColumnNextgenAVX512                 313.90%      1.34s  747.56m
mulDoubleSameColumnNextgenAllOpt                 308.22%      1.36s  734.02m
mulDoubleSameColumnCompile                                  45.74ms    21.86
----------------------------------------------------------------------------
mulDoubleConstantVelox_________Base                           3.04s  329.12m
mulDoubleConstantNextGen                          65.10%      4.67s  214.25m
mulDoubleConstantNextGen_______Base                           4.64s  215.43m
mulDoubleConstantNextgenAVX2                     276.22%      1.68s  595.07m
mulDoubleConstantNextgenAVX512                   280.37%      1.66s  604.02m
mulDoubleConstantNextgenAllOpt                   279.72%      1.66s  602.62m
mulDoubleConstantCompile                                    46.10ms    21.69
----------------------------------------------------------------------------
mulDoubleNestedVelox_________Base                             4.55s  219.72m
mulDoubleNestedNextGen                            93.10%      4.89s  204.57m
mulDoubleNestedNextGen_______Base                             4.95s  202.14m
mulDoubleNestedNextgenAVX2                       313.40%      1.58s  633.51m
mulDoubleNestedNextgenAVX512                     325.82%      1.52s  658.62m
mulDoubleNestedNextgenAllOpt                     301.88%      1.64s  610.22m
mulDoubleNestedCompile                                      46.08ms    21.70
----------------------------------------------------------------------------
mulDoubleNestedDeepVelox_________Base                         9.26s  108.00m
mulDoubleNestedDeepNextGen                       192.79%      4.80s  208.21m
mulDoubleNestedDeepNextGen_______Base                         4.83s  207.11m
mulDoubleNestedDeepNextgenAVX2                   284.61%      1.70s  589.44m
mulDoubleNestedDeepNextgenAVX512                 275.56%      1.75s  570.70m
mulDoubleNestedDeepNextgenAllOpt                 295.89%      1.63s  612.80m
mulDoubleNestedDeepCompile                                  46.42ms    21.54
----------------------------------------------------------------------------
PlusCheckedVeloxI64                                           4.16s  240.42m
plusI64Velox_________Base                                     3.30s  302.61m
plusI64NextGen                                    80.00%      4.13s  242.08m
plusI64NextGen_______Base                                     4.13s  242.39m
plusI64NextgenAVX2                               238.91%      1.73s  579.10m
plusI64NextgenAVX512                             318.55%      1.30s  772.12m
plusI64NextgenAllOpt                             329.08%      1.25s  797.65m
plusI64Compile                                              45.79ms    21.84
----------------------------------------------------------------------------
mulAndAddVelox_________Base                                  15.96s   62.65m
mulAndAddNextGen                                 322.63%      4.95s  202.13m
mulAndAddNextGen_______Base                                   4.96s  201.54m
mulAndAddNextgenAVX2                              99.68%      4.98s  200.90m
mulAndAddNextgenAVX512                           100.87%      4.92s  203.30m
mulAndAddNextgenAllOpt                            80.23%      6.18s  161.71m
mulAndAddCompile                                            46.33ms    21.58
----------------------------------------------------------------------------
eqVelox_________Base                                          3.76s  265.93m
eqNextGen                                         73.62%      5.11s  195.77m
eqNextGen_______Base                                          5.12s  195.46m
eqNextgenAVX2                                     99.35%      5.15s  194.19m
eqNextgenAVX512                                  100.12%      5.11s  195.70m
eqNextgenAllOpt                                   53.83%      9.50s  105.21m
eqCompile                                                   46.47ms    21.52
----------------------------------------------------------------------------
eqConstantVelox_________Base                                  3.77s  265.18m
eqConstantNextGen                                 70.25%      5.37s  186.29m
eqConstantNextGen_______Base                                  5.12s  195.49m
eqConstantNextgenAVX2                             99.03%      5.17s  193.60m
eqConstantNextgenAVX512                          100.12%      5.11s  195.73m
eqConstantNextgenAllOpt                           53.95%      9.48s  105.46m
eqConstantCompile                                           46.55ms    21.48
----------------------------------------------------------------------------
eqBoolVelox_________Base                                      4.34s  230.53m
eqBoolNextGen                                     75.51%      5.74s  174.08m
eqBoolNextGen_______Base                                      5.76s  173.49m
eqBoolNextgenAVX2                                100.07%      5.76s  173.61m
eqBoolNextgenAVX512                              100.45%      5.74s  174.28m
eqBoolNextgenAllOpt                               61.02%      9.45s  105.86m
eqBoolCompile                                               46.98ms    21.29
----------------------------------------------------------------------------
neqVelox_________Base                                         3.58s  279.48m
neqNextGen                                        70.46%      5.08s  196.93m
neqNextGen_______Base                                         5.10s  195.99m
neqNextgenAVX2                                   101.08%      5.05s  198.12m
neqNextgenAVX512                                  98.31%      5.19s  192.68m
neqNextgenAllOpt                                  58.33%      8.75s  114.32m
neqCompile                                                  46.11ms    21.69
----------------------------------------------------------------------------
gtVelox_________Base                                          3.64s  274.67m
gtNextGen                                         67.84%      5.37s  186.34m
gtNextGen_______Base                                          5.30s  188.75m
gtNextgenAVX2                                     99.32%      5.33s  187.46m
gtNextgenAVX512                                   99.70%      5.31s  188.18m
gtNextgenAllOpt                                   60.92%      8.70s  114.98m
gtCompile                                                   46.12ms    21.68
----------------------------------------------------------------------------
ltVelox_________Base                                          3.50s  285.31m
ltNextGen                                         65.73%      5.33s  187.52m
ltNextGen_______Base                                          5.31s  188.30m
ltNextgenAVX2                                     99.65%      5.33s  187.64m
ltNextgenAVX512                                   99.21%      5.35s  186.81m
ltNextgenAllOpt                                   61.31%      8.66s  115.45m
ltCompile                                                   46.20ms    21.65
----------------------------------------------------------------------------
andVelox_________Base                                        34.02s   29.40m
andNextGen                                       386.49%      8.80s  113.62m
andNextGen_______Base                                         8.81s  113.53m
andNextgenAVX2                                   110.83%      7.95s  125.82m
andNextgenAVX512                                 113.58%      7.75s  128.95m
andNextgenAllOpt                                  83.98%     10.49s   95.33m
andCompile                                                  47.22ms    21.18
----------------------------------------------------------------------------
orVelox_________Base                                         35.14s   28.45m
orNextGen                                        453.80%      7.74s  129.13m
orNextGen_______Base                                          7.74s  129.20m
orNextgenAVX2                                    115.02%      6.73s  148.60m
orNextgenAVX512                                  113.24%      6.83s  146.31m
orNextgenAllOpt                                   75.68%     10.23s   97.78m
orCompile                                                   47.09ms    21.24
----------------------------------------------------------------------------
conjunctsNestedVelox_________Base                           2.71min    6.15m
conjunctsNestedNextGen                           726.58%     22.40s   44.65m
conjunctsNestedNextGen_______Base                            22.81s   43.84m
conjunctsNestedNextgenAVX2                       135.79%     16.80s   59.53m
conjunctsNestedNextgenAVX512                     133.88%     17.04s   58.70m
conjunctsNestedNextgenAllOpt                     120.11%     18.99s   52.66m
conjunctsNestedCompile                                      49.28ms    20.29
----------------------------------------------------------------------------
```

### Why are the changes needed?
performance optimization

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UTs

### Which label does this PR belong to?
INFRA

### CODE REVIEW

File: cpp/src/cider/exec/nextgen/context/CodegenContext.h [hunk 1]
```diff
@@ -72,7 +72,7 @@ struct CodegenOptions {
   bool check_bit_vector_clear_opt = false;
   bool set_null_bit_vector_opt = false;
   bool branchless_logic = true;
-  bool enable_vectorize = false;
+  bool enable_vectorize = true;
```
Comment (yma11) [inline cpp/src/cider/exec/nextgen/context/CodegenContext.h:75]: I think some of the code path in codegen will be disabled if we enable vectorize by default. Should we cover both of the options?

File: cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h [hunk 2]
```diff
@@ -31,9 +31,9 @@ struct CompilationOptions {
   bool optimize_ir = true;
   bool aggressive_jit_compile = true;
   bool dump_ir = false;
-  bool enable_vectorize = false;
+  bool enable_vectorize = true;
   bool enable_avx2 = true;
-  bool enable_avx512 = false;
+  bool enable_avx512 = true;
```
Comment (bigPYJ1151) [inline cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h:36]: Maybe AVX512 shouldn't be enabled, seems it has no benefit to performance now.

File: cpp/src/cider/exec/nextgen/context/CodegenContext.h (hunk 1 above)
Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/context/CodegenContext.h:75]: any example? I hope existing uts can approve the correctness.

File: cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h (hunk 2 above)
Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h:36]: I'll try to test on icx, if nothing gains, I'll keep it closed.

Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h:36]: In IceLake, it has no gains. Not sure about SPR.

File: cpp/src/cider/exec/nextgen/context/CodegenContext.h (hunk 1 above)
Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/context/CodegenContext.h:75]: seems some test will failed, 
```bash
The following tests FAILED:
         27 - CiderNextgenQueryRunnerTest (Failed)
         36 - CaseWhenTest (SEGFAULT)
         37 - CastTest (SEGFAULT)
         38 - CiderMathOpNextgenTest (Failed)
         39 - DateTest (Failed)
         40 - FilterTest (SEGFAULT)
         42 - PrimitiveTypeQueryNextgenTest (Failed)
         43 - StringTest (SEGFAULT)
```

Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/context/CodegenContext.h:75]: open `enable_vectorize`, all uts can pass. but some failed after removing `memset`.

Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/context/CodegenContext.h:75]: > open `enable_vectorize`, all uts can pass. but some failed after removing `memset`.

some test with variable column failed caused by offset[0] not init after removing `memset`, fixed.

File: cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h (hunk 2 above)
Comment (Ma-Jian1) [reply cpp/src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h]: no performance gains for the small batch. for large batch,  it depends. 
keep it closed

**Comment by bigPYJ1151** [comment]
There are ~7000 code lines added, seems some undesired files were committed.

**Review by bigPYJ1151 — APPROVED** [review]

File: cpp/src/cider/exec/module/batch/CiderArrowBufferHolder.cpp [hunk 3]
```diff
@@ -63,7 +63,6 @@ void CiderArrowArrayBufferHolder::allocBuffer(size_t index, size_t bytes) {
     }
   } else {
     buffers_[index] = allocator_->allocate(bytes);
-    memset(buffers_[index], 0, bytes);  // set new initialized bytes to 0
```
Comment (jikunshang) [inline cpp/src/cider/exec/module/batch/CiderArrowBufferHolder.cpp:66]: I think it's more safe to keep memset here. any performance gain if removed this?

Comment (Ma-Jian1) [reply cpp/src/cider/exec/module/batch/CiderArrowBufferHolder.cpp:66]: about 50% was reduced from `allocBuffer`, which is about 27% in `query_func`.

**Review by yma11 — APPROVED** [review]

**Pull request merged**